src/web_worker_kernel.ts

```diff
@@ -40,27 +29,83 @@ export class WebWorkerKernel implements IKernel {
    * @param options The instantiation options for a new WebWorkerKernel
    */
   constructor(options: WebWorkerKernel.IOptions) {
-    const { id, name, sendMessage, location, kernelspec, contentsManager } =
+    const { id, name, sendMessage, location, kernelSpec, contentsManager } =
       options;
     this._id = id;
     this._name = name;
     this._location = location;
-    this._kernelspec = kernelspec;
+    this._kernelSpec = kernelSpec;
     this._contentsManager = contentsManager;
     this._sendMessage = sendMessage;
-    this._worker = new Worker(new URL('./worker.js', import.meta.url), {
-      type: 'module'
-    });
-
-    this._worker.onmessage = this._processWorkerMessage.bind(this);
+    this._worker = this.initWorker(options);
+    this._remoteKernel = this.initRemote(options);
 
-    this._remote = coincident(this._worker) as IXeusKernel;
-
-    this.setupFilesystemAPIs();
+    this.initFileSystem(options);
+  }
 
-    this._remote.initialize(this._kernelspec, PageConfig.getBaseUrl());
+  /**
+   * Load the worker.
+   */
+  protected initWorker(options: WebWorkerKernel.IOptions): Worker {
+    if (crossOriginIsolated) {
+      return new Worker(new URL('./coincident.worker.js', import.meta.url), {
+        type: 'module'
+      });
+    } else {
+      return new Worker(new URL('./comlink.worker.js', import.meta.url), {
+        type: 'module'
+      });
+    }
+  }
 
-    this.initFileSystem(options);
+  /**
+   * Initialize the remote kernel.
+   * Use coincident if crossOriginIsolated, comlink otherwise
+   * See the two following issues for more context:
+   *  - https://github.com/jupyterlite/jupyterlite/issues/1424
+   *  - https://github.com/jupyterlite/xeus/issues/102
+   */
+  protected initRemote(
+    options: WebWorkerKernel.IOptions
+  ): IXeusWorkerKernel | Remote<IXeusWorkerKernel> {
+    let remote: IXeusWorkerKernel | Remote<IXeusWorkerKernel>;
+    if (crossOriginIsolated) {
+      this._worker.onmessage = this._processCoincidentWorkerMessage.bind(this);
```

Comment:
OK that would confirm the observation in the comment above I guess? https://github.com/jupyterlite/xeus/pull/108#issuecomment-2293505406

Reply:
Ideally we would have had the same approach for both `jupyterlite-pyodide-kernel` and `jupyterlite-xeus`, so it's easier when switching code bases.

But then it means we would need a way for the xeus kernel to first send message to the worker context without the use of `postMessage`, and then implement the same logic as in `jupyterlite-pyodide-kernel` to handle the main thread <-> worker communication with either `comlink` or `coincident`.

Reply:
@martinRenou maybe we can leave a comment next to this line to reference the discussion on this PR, for future reference, and in case we wonder in the future why it is done this way in `jupyterlite-xeus`.